RAT/project.py

```diff
@@ -170,5 +207,43 @@ def check_allowed_values(self, attribute: str, field_list: list[str], allowed_va
             for field in field_list:
                 value = getattr(model, field)
                 if value and value not in allowed_values:
-                    setattr(model, field, '')
-                    raise ValueError(f'The parameter "{value}" has not been defined in the list of allowed values.')
+                    raise ValueError(f'The value "{value}" in the "{field}" field of "{attribute}" must be defined in '
+                                     f'"{values_defined_in[attribute + "." + field]}".')
+
+    def _classlist_wrapper(self, class_list: 'ClassList', func: Callable):
+        """Defines the function used to wrap around ClassList routines to force revalidation.
+
+        Parameters
+        ----------
+        class_list : ClassList
+            The ClassList defined in the "Project" model that is being modified.
+        func : Callable
+            The routine being wrapped.
+
+        Returns
+        -------
+        wrapped_func : Callable
+            The wrapped routine.
+        """
+        @functools.wraps(func)
+        def wrapped_func(*args, **kwargs):
+            """Run the given function and then revalidate the "Project" model. If any exception is raised, restore
+            the previous state of the given ClassList and report details of the exception.
+            """
+            previous_state = copy.deepcopy(getattr(class_list, 'data'))
```

Comment:
Is a deep copy needed, I was thinking a shallow copy would work?

Reply:
No, a deep copy is required to copy the contents of the ClassList correctly.

Reply:
A shallow copy will have the order of the list but reference to the data, I thought the order of the list is what was needed. What happens when you shallow copy?

Reply:
If you shallow copy, you have a reference to the data. The code then changes the data, so if the data is changed in a way that proves invalid, the copy has a reference to the same changed data, and cannot restore the old version of the data as desired. The tests cover this case, simply changing "deepcopy" to "copy" will cause them to fail.